src/inmanta/env.py

```diff
@@ -424,94 +389,96 @@ def compose_list_command(
-            if sys.platform == "win32"
-            else os.path.join(env_path, "bin", python_name)
-        )
-
-    @classmethod
-    def get_site_dir_for_env_path(cls, env_path: str) -> str:
-        """
-        Return the site directory for a given venv directory.
-        """
-        return (
-            os.path.join(env_path, "Lib", "site-packages")
-            if sys.platform == "win32"
-            else os.path.join(
-                env_path, "lib", "python%s" % ".".join(str(digit) for digit in sys.version_info[:2]), "site-packages"
-            )
-        )
-
-    def _run_pip_install_command(
-        self,
+    def run_pip_install_command_from_config(
+        cls,
         python_path: str,
-        requirements: Optional[List[Requirement]] = None,
-        paths: Optional[List[LocalPackagePath]] = None,
-        index_urls: Optional[List[str]] = None,
+        config: PipConfig,
+        requirements: Optional[Sequence[Requirement]] = None,
+        requirements_files: Optional[List[str]] = None,
         upgrade: bool = False,
         upgrade_strategy: PipUpgradeStrategy = PipUpgradeStrategy.ONLY_IF_NEEDED,
-        allow_pre_releases: bool = False,
         constraints_files: Optional[List[str]] = None,
-        requirements_files: Optional[List[str]] = None,
-        use_pip_config: Optional[bool] = False,
+        paths: Optional[List[LocalPackagePath]] = None,
     ) -> None:
-        cmd: List[str] = PipCommandBuilder.compose_install_command(
-            python_path=python_path,
-            requirements=requirements,
-            paths=paths,
-            index_urls=index_urls,
-            upgrade=upgrade,
-            upgrade_strategy=upgrade_strategy,
-            allow_pre_releases=allow_pre_releases,
-            constraints_files=constraints_files,
-            requirements_files=requirements_files,
-            use_pip_config=use_pip_config,
+        # What
+        requirements = requirements if requirements is not None else []
+        requirements_files = requirements_files if requirements_files is not None else []
+
+        paths = paths if paths is not None else []
+        local_paths: Iterator[LocalPackagePath] = (
+            # make sure we only try to install from a local source: add leading `./` and trailing `/` to explicitly tell pip
+            # we're pointing to a local directory.
+            LocalPackagePath(path=os.path.join(".", path.path, ""), editable=path.editable)
+            for path in paths
         )
+
+        install_args = [
+            *(str(requirement) for requirement in requirements),
+            *chain.from_iterable(["-r", f] for f in requirements_files),
+            *chain.from_iterable(["-e", path.path] if path.editable else [path.path] for path in local_paths),
+        ]
+
+        # From where
+        index_args: list[str] = []
+        if config.index_url:
+            index_args.append("--index-url")
+            index_args.append(config.index_url)
```

Comment:
If the config doesn't set index url and we are not using pip config, do I need to set `--no-index-url` ?

Reply:
Yes

Reply:
Though in practice we always raise an error in that case, don't we?

Reply:
not if you also set paths

Reply:
added